firebase-firestore/src/main/java/com/google/firebase/firestore/util/GrpcCallProvider.java

```diff
@@ -0,0 +1,224 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.util;
+
+import android.content.Context;
+import com.google.android.gms.common.GooglePlayServicesNotAvailableException;
+import com.google.android.gms.common.GooglePlayServicesRepairableException;
+import com.google.android.gms.security.ProviderInstaller;
+import com.google.firestore.v1.FirestoreGrpc;
+import io.grpc.CallCredentials;
+import io.grpc.CallOptions;
+import io.grpc.ClientCall;
+import io.grpc.ManagedChannel;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Manages the GRPC channel and encapsulates all SSL and GRPC initialization.
+ *
+ * <p>All operations are dispatched to an internal worker queue and are not executed until the SSL
+ * and GRPC Stub initialization completes.
+ */
+// PORTING NOTE: This class only exists on Android.
+class GrpcCallProvider {
+
+  private static final String LOG_TAG = "GrpcCallProvider";
+
+  private static final RejectedExecutionHandler IGNORE_REJECTIONS_HANDLER =
+      (runnable, executor) -> {
+        // We ignore rejected executions, since some GRPC messages can arrive after shutdown.
+      };
+
+  private final ExecutorService channelQueue;
+  private final ManagedChannel channel;
+
+  private CallOptions callOptions;
+  private boolean shutdown = false;
+
+  GrpcCallProvider(
+      AsyncQueue queue,
+      Context context,
+      ManagedChannel grpcChannel,
+      CallCredentials firestoreHeaders) {
+    this.channel = grpcChannel;
+
+    ThreadFactory threadFactory =
+        runnable -> {
+          Thread thread = Executors.defaultThreadFactory().newThread(runnable);
+          thread.setName("FirestoreGrpcWorker");
+          thread.setDaemon(true);
+          thread.setUncaughtExceptionHandler((crashingThread, throwable) -> queue.panic(throwable));
+          return thread;
+        };
+
+    this.channelQueue =
+        new ThreadPoolExecutor(
+            /* corePoolSize= */ 1,
+            /* maximumPoolSize= */ 1,
+            /* keepAliveTime= */ 0L,
+            TimeUnit.MILLISECONDS,
+            new LinkedBlockingQueue<>(),
+            threadFactory,
+            IGNORE_REJECTIONS_HANDLER);
+
+    // We execute network initialization on an internal queue to not block operations that
+    // depend on the AsyncQueue.
+    channelQueue.execute(
+        () -> {
+          FirestoreGrpc.FirestoreStub firestoreStub =
```

Comment:
This seems like the reverse order to what came before (we did the provider installer stuff first in the initialize). Is this intended?